contracts/contracts/coordination/Coordinator.sol

```diff
@@ -525,6 +560,135 @@ contract Coordinator is Initializable, AccessControlDefaultAdminRulesUpgradeable
         processReimbursement(initialGasLeft);
     }
 
+    function handoverRequest(
+        uint32 ritualId,
+        address departingParticipant,
+        address incomingParticipant
+    ) external onlyRole(HANDOVER_SUPERVISOR_ROLE) {
+        require(isRitualActive(ritualId), "Ritual is not active");
+        require(
+            isParticipant(ritualId, departingParticipant),
+            "Departing node must be a participant"
+        );
+        require(
+            !isParticipant(ritualId, incomingParticipant),
+            "Incoming node cannot be a participant"
+        );
+
+        Handover storage handover = handovers[getHandoverKey(ritualId, departingParticipant)];
+        HandoverState state = getHandoverState(handover);
+
+        require(
+            state == HandoverState.NON_INITIATED || state == HandoverState.HANDOVER_TIMEOUT,
+            "Handover already requested"
+        );
+        require(isProviderKeySet(incomingParticipant), "Incoming provider has not set public key");
+        handover.handoverRequestTimestamp = uint32(block.timestamp);
+        handover.incomingProvider = incomingParticipant;
+        emit HandoverRequest(ritualId, departingParticipant, incomingParticipant);
+    }
+
+    function postBlindedShare(uint32 ritualId, bytes calldata blindedShare) external {
+        require(isRitualActive(ritualId), "Ritual is not active");
+
+        address provider = application.operatorToStakingProvider(msg.sender);
+        Handover storage handover = handovers[getHandoverKey(ritualId, provider)];
+        require(
+            getHandoverState(handover) == HandoverState.HANDOVER_AWAITING_BLIND_SHARE,
+            "Not waiting for blind share"
+        );
+
+        handover.handoverBlindedShare = blindedShare;
+        emit BlindedSharePosted(ritualId, provider);
+    }
+
+    function cancelHandover(
+        uint32 ritualId,
+        address departingParticipant
+    ) external onlyRole(HANDOVER_SUPERVISOR_ROLE) {
+        Handover storage handover = handovers[getHandoverKey(ritualId, departingParticipant)];
+        address incomingParticipant = handover.incomingProvider;
+
+        require(
+            getHandoverState(handover) != HandoverState.NON_INITIATED,
+            "Handover not requested"
+        );
+        handover.handoverRequestTimestamp = 0;
+        handover.incomingProvider = address(0);
+        delete handover.handoverBlindedShare;
+
+        emit HandoverCanceled(ritualId, departingParticipant, incomingParticipant);
+    }
+
+    function finalizeHandover(
+        uint32 ritualId,
+        address departingParticipant
+    ) external onlyRole(HANDOVER_SUPERVISOR_ROLE) {
```

Comment:
For contract size purposes, does it make sense to add a `bool successful` parameter here and then generalize the method so that we can remove `cancelHandover()?`

Reply:
it can save very tiny amount of contract size, I'd not do it now because it will decrease readability